spring-integration-amqp/src/main/java/org/springframework/integration/amqp/outbound/AsyncOutboundGateway.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.integration.amqp.outbound;
+
+import org.springframework.amqp.core.AmqpMessageReturnedException;
+import org.springframework.amqp.rabbit.core.AsyncRabbitTemplate;
+import org.springframework.amqp.rabbit.core.AsyncRabbitTemplate.RabbitMessageFuture;
+import org.springframework.amqp.rabbit.support.CorrelationData;
+import org.springframework.amqp.support.converter.MessageConverter;
+import org.springframework.amqp.support.converter.SimpleMessageConverter;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageHandlingException;
+import org.springframework.messaging.MessagingException;
+import org.springframework.messaging.support.ErrorMessage;
+import org.springframework.util.concurrent.ListenableFutureCallback;
+
+import reactor.core.support.Assert;
+
+/**
+ * An outbound gateway where the sending thread is released immediately and the reply
+ * is sent on the async template's listener container thread.
+ *
+ * @author Gary Russell
+ * @since 4.3
+ *
+ */
+public class AsyncOutboundGateway extends AbstractAmqpOutboundEndpoint {
+
+	private final AsyncRabbitTemplate template;
+
+	private MessageConverter messageConverter = new SimpleMessageConverter();
+
+	public AsyncOutboundGateway(AsyncRabbitTemplate template) {
+		Assert.notNull(template, "AsyncRabbitTemplate cannot be null");
+		this.template = template;
+	}
+
+	public void setMessageConverter(MessageConverter messageConverter) {
+		this.messageConverter = messageConverter;
+	}
+
+	@Override
+	public String getComponentType() {
+		return "amqp:outbound-async-gateway";
+	}
+
+	@Override
+	protected void endpointInit() {
+		super.endpointInit();
+		setAsyncReplySupported(true);
+		// TODO - add getConnectionFactory() to async template.
+		// this.setConnectionFactory(template.getConnectionFactory());
+	}
+
+	@Override
+	protected Object handleRequestMessage(Message<?> requestMessage) {
+		Assert.state(getAsyncReplySupported(), "Gateway not initialized - did you call afterPropertiesSet() ?");
```

Comment:
Not sure why we can't do `setAsyncReplySupported(true)` from the ctor and remove this assert...